File: src/test/java/org/cloudfoundry/promregator/discovery/CFMultiDiscovererTest.java
package org.cloudfoundry.promregator.discovery;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

import java.time.Clock;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import org.cloudfoundry.promregator.JUnitTestUtils;
import org.cloudfoundry.promregator.cfaccessor.CFAccessorMock;
import org.cloudfoundry.promregator.config.Target;
import org.cloudfoundry.promregator.messagebus.MessageBus;
import org.cloudfoundry.promregator.messagebus.MessageBusTopic;
import org.cloudfoundry.promregator.messagebus.MessageSubscriber;
import org.cloudfoundry.promregator.scanner.Instance;
import org.cloudfoundry.promregator.scanner.ResolvedTarget;
import org.cloudfoundry.promregator.scanner.TargetResolver;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = CFDiscovererTestSpringApplication.class)
@TestPropertySource(locations="default.properties")
public class CFMultiDiscovererTest implements MessageSubscriber {
	
	@AfterAll
	public static void cleanUp() {
		JUnitTestUtils.cleanUpAll();
	}
	
	@Autowired
	private CFMultiDiscoverer cfDiscoverer;

	@Autowired
	private TargetResolver targetResolver;

	@Autowired
	private Clock clock;
	
	@Autowired
	private MessageBus messageBus;
	
	private List<Instance> removerTriggerForInstances = new LinkedList<>();
	
	@Override
	public void receiveMessage(String topic, Object message) {
		Assertions.assertEquals(MessageBusTopic.DISCOVERER_INSTANCE_REMOVED, topic);
		
		Instance instance = (Instance) message;
		
		this.removerTriggerForInstances.add(instance);
	}

	@Test
	void testDiscoverWithCleanup() throws InterruptedException {
		this.messageBus.subscribe(MessageBusTopic.DISCOVERER_INSTANCE_REMOVED, this);
		
		List<ResolvedTarget> resolvedTargets = new ArrayList<>();
		ResolvedTarget aTarget = new ResolvedTarget();
		aTarget.setOrgName("unittestorg");
		aTarget.setSpaceName("unittestspace");
		aTarget.setApplicationName("testapp");
		aTarget.setApplicationId(CFAccessorMock.UNITTEST_APP1_UUID);
		aTarget.setProtocol("https");
		aTarget.setPath("/metrics");
		aTarget.setOriginalTarget(new Target());
		resolvedTargets.add(aTarget);
		when(targetResolver.resolveTargets(any())).thenReturn(resolvedTargets);

		List<Instance> result = this.cfDiscoverer.discover(null, null);
		
		Assertions.assertEquals(2, result.size());
		
		boolean testapp1_instance1 = false;
		boolean testapp1_instance2 = false;
		
		Instance i1 = null;
		Instance i2 = null;
		
		for (Instance instance : result) {
			String instanceId = instance.getInstanceId();
			
			if (instanceId.equals(CFAccessorMock.UNITTEST_APP1_UUID+":0")) {
				testapp1_instance1 = true;
				Assertions.assertEquals("https://hostapp1.shared.domain.example.org/metrics", instance.getAccessUrl());
				i1 = instance;
			} else if (instanceId.equals(CFAccessorMock.UNITTEST_APP1_UUID+":1")) {
				testapp1_instance2 = true;
				Assertions.assertEquals("https://hostapp1.shared.domain.example.org/metrics", instance.getAccessUrl());
				i2 = instance;
			} else if (instanceId.equals(CFAccessorMock.UNITTEST_APP2_UUID+":0")) {
				Assertions.fail("Should not have been returned");
			}
		}
		Assertions.assertTrue(testapp1_instance1);
		Assertions.assertTrue(testapp1_instance2);
		
		Assertions.assertNotNull(i1);
		Assertions.assertNotNull(i2);
		
		Assertions.assertTrue(this.cfDiscoverer.isInstanceRegistered(i1));
		Assertions.assertTrue(this.cfDiscoverer.isInstanceRegistered(i2));
		Assertions.assertTrue(this.removerTriggerForInstances.isEmpty());
		
		// early cleaning does not change anything
		this.cfDiscoverer.cleanup();
		Assertions.assertTrue(this.cfDiscoverer.isInstanceRegistered(i1));
		Assertions.assertTrue(this.cfDiscoverer.isInstanceRegistered(i2));
		
		// Wait a little to allow JMX do its job... (if it really did something)
		for (int i = 0;i<10;i++) {
			if (!this.removerTriggerForInstances.isEmpty()) {
				Thread.sleep(100);
				continue;
			}
			Assertions.assertTrue(this.removerTriggerForInstances.isEmpty());
		}
		
		// later cleaning does...
		this.cfDiscoverer.setClock(Clock.offset(this.clock, Duration.ofMinutes(10)));
		
		this.cfDiscoverer.cleanup();
		
		Assertions.assertFalse(this.cfDiscoverer.isInstanceRegistered(i1));
		Assertions.assertFalse(this.cfDiscoverer.isInstanceRegistered(i2));
		
		for (int i = 0;i<10;i++) {
			if (this.removerTriggerForInstances.isEmpty()) {
				Thread.sleep(400);
				continue;
			}
			Assertions.assertEquals(2, this.removerTriggerForInstances.size());
		}
	}


}

File: src/test/java/org/cloudfoundry/promregator/discovery/CFMultiDiscovererTest.kt
package org.cloudfoundry.promregator.discovery

import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.runs
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.cloudfoundry.promregator.cfaccessor.CFAccessorMock
import org.cloudfoundry.promregator.config.PromregatorConfiguration
import org.cloudfoundry.promregator.config.Target
import org.cloudfoundry.promregator.messagebus.MessageBusDestination
import org.cloudfoundry.promregator.scanner.AppInstanceScanner
import org.cloudfoundry.promregator.scanner.Instance
import org.cloudfoundry.promregator.scanner.ResolvedTarget
import org.cloudfoundry.promregator.scanner.TargetResolver
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.fail
import org.springframework.jms.core.JmsTemplate
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono
import java.time.Clock
import java.time.Instant
import java.time.temporal.ChronoUnit

class CFMultiDiscovererTest {

    private lateinit var appInstanceScanner: AppInstanceScanner
    private lateinit var cfDiscoverer: CFMultiDiscoverer
    private lateinit var  targetResolver: TargetResolver
    private lateinit var  clock: Clock
    private val promConfig = PromregatorConfiguration()
    private lateinit var jmsTemplate : JmsTemplate

    @BeforeEach
    fun setup() {
        targetResolver = mockk()
        appInstanceScanner = mockk()
        jmsTemplate = mockk()
        clock = mockk()
        every { clock.instant() } returns Instant.parse("2007-12-03T10:15:30.00Z")

        cfDiscoverer = CFMultiDiscoverer(targetResolver, appInstanceScanner, promConfig, jmsTemplate, clock)
    }

    @Test
    fun `discovery should register instaces and cleanup should remove them`() {
        val resolvedTarget = ResolvedTarget().apply {
            orgName = "unittestorg"
            spaceName = "unittestspace"
            applicationName = "testapp"
            applicationId = CFAccessorMock.UNITTEST_APP1_UUID
            protocol = "https"
            path = "/metrics"
            originalTarget = Target()
        }
        val resolvedTargets = listOf(resolvedTarget)

        every { jmsTemplate.convertAndSend(MessageBusDestination.DISCOVERER_INSTANCE_REMOVED, any<String>()) } just runs
        every { targetResolver.resolveTargets(any()) } returns resolvedTargets.toMono()

        every {appInstanceScanner.determineInstancesFromTargets(any(), null, null)} returns listOf(
                Instance(resolvedTarget, CFAccessorMock.UNITTEST_APP1_UUID +":0", "http://url123"),
                Instance(resolvedTarget, CFAccessorMock.UNITTEST_APP1_UUID +":1", "http://url123")
        ).toMono()

        val result = cfDiscoverer.discover(null, null)

        assertThat(result).hasSize(2)


        val i1 = result.find { it.instanceId == CFAccessorMock.UNITTEST_APP1_UUID + ":0"} ?: fail("Instance 0 wasn't found")
        val i2 = result.find { it.instanceId == CFAccessorMock.UNITTEST_APP1_UUID + ":1"} ?: fail("Instance 1 wasn't found")

        assertThat(cfDiscoverer.isInstanceRegistered(i1)).isTrue()
        assertThat(cfDiscoverer.isInstanceRegistered(i2)).isTrue()

        verify(exactly = 0) { jmsTemplate.convertAndSend(any<String>(), any<String>())}

        // early cleaning does not change anything
        cfDiscoverer.cleanup()
        assertThat(cfDiscoverer.isInstanceRegistered(i1)).isTrue()
        assertThat(cfDiscoverer.isInstanceRegistered(i2)).isTrue()
        verify(exactly = 0) { jmsTemplate.convertAndSend(any<String>(), any<String>())}

        // later cleaning does...
        every { clock.instant() } returns Instant.parse("2007-12-03T10:15:30.00Z").plus(10, ChronoUnit.MINUTES)

        cfDiscoverer.cleanup()
        assertThat(cfDiscoverer.isInstanceRegistered(i1)).isFalse()
        assertThat(cfDiscoverer.isInstanceRegistered(i2)).isFalse()
        verify(exactly = 2) { jmsTemplate.convertAndSend(any<String>(), any<String>())}
    }

}